### PR TITLE
DLP: Added sample for reidentify content using fpe

### PR DESCRIPTION
**Pull request by jainamtrivedi-crest**
- Added sample for [Reidentify Content Using FPE](https://cloud.google.com/dlp/docs/samples/dlp-reidentify-fpe) and test cases

### CODE REVIEW

File: dlp/api/Snippets.Tests/ReidentifyContentUsingFpeTests.cs [hunk 1]
```diff
@@ -0,0 +1,35 @@
...
+        private DlpTestFixture _fixture;
+        public ReidentifyContentUsingFpeTests(DlpTestFixture fixture) => _fixture = fixture;
+
+        [Fact]
+        public void TestReidentifyWithFpe()
+        {
+            var finding = "731997681";
+            var text = $"My SSN is SSN_TOKEN(9):{finding}";
+            var expectedTransformFake = $"My SSN is {finding}";
+            var result = ReidentifyContentUsingFpe.ReidentifyContent(_fixture.ProjectId, text, _fixture.KeyName, _fixture.WrappedKey);
```
Comment (micheldavid) [inline dlp/api/Snippets.Tests/ReidentifyContentUsingFpeTests.cs:30]: avoid using "var" when the type is not easy to infer from surrounding context

File: dlp/api/Snippets/ReidentifyContentUsingFpe.cs [hunk 2]
```diff
@@ -0,0 +1,106 @@
...
+        var request = new ReidentifyContentRequest
+        {
+            ParentAsLocationName = new LocationName(projectId, "global"),
+            InspectConfig = inspectConfig,
+            ReidentifyConfig = reidentifyConfig,
+            Item = new ContentItem { Value = text },
+        };
+
+        // Call the API.
+        var response = dlp.ReidentifyContent(request);
```
Comment (micheldavid) [inline dlp/api/Snippets/ReidentifyContentUsingFpe.cs:97]: same here, abour "var"

File: dlp/api/Snippets.Tests/ReidentifyContentUsingFpeTests.cs [hunk 3]
```diff
@@ -0,0 +1,35 @@
...
+    {
+        private DlpTestFixture _fixture;
+        public ReidentifyContentUsingFpeTests(DlpTestFixture fixture) => _fixture = fixture;
+
+        [Fact]
+        public void TestReidentifyWithFpe()
+        {
+            var finding = "731997681";
+            var text = $"My SSN is SSN_TOKEN(9):{finding}";
+            var expectedTransformFake = $"My SSN is {finding}";
```
Comment (micheldavid) [inline dlp/api/Snippets.Tests/ReidentifyContentUsingFpeTests.cs:29]: not really a fake, but a guide of how many characters to expect... i'd recommend storing the expected length instead, and use that to compare below.

`var expectedValueLength = "My SSN is ".length + finding.length;`

**Review by micheldavid — APPROVED** [review]

File: dlp/api/Snippets.Tests/ReidentifyContentUsingFpeTests.cs (hunk 1 above)
Comment (jainamtrivedi-crest) [reply dlp/api/Snippets.Tests/ReidentifyContentUsingFpeTests.cs:30]: In the other samples (pre-written), only "var" has been used. So, just to be consistent; I have also used "var" at places.

File: dlp/api/Snippets.Tests/ReidentifyContentUsingFpeTests.cs (hunk 3 above)
Comment (jainamtrivedi-crest) [reply dlp/api/Snippets.Tests/ReidentifyContentUsingFpeTests.cs]: Done. Thanks

**Review by amanda-tarafa — CHANGES_REQUESTED** [review]
Tests are failing

```csharp
Failed GoogleCloudSamples.ReidentifyContentUsingFpeTests.TestReidentifyWithFpe [3 ms]
  Error Message:
   System.ArgumentNullException : Value cannot be null. (Parameter 'value')
  Stack Trace:
     at Google.Protobuf.ProtoPreconditions.CheckNotNull[T](T value, String name)
   at Google.Cloud.Dlp.V2.KmsWrappedCryptoKey.set_CryptoKeyName(String value)
   at ReidentifyContentUsingFpe.ReidentifyContent(String projectId, String text, String keyName, String wrappedKey, FfxCommonNativeAlphabet alphabet, InfoType surrogateType) in /tmpfs/src/github/dotnet-docs-samples/dlp/api/Snippets/ReidentifyContentUsingFpe.cs:line 41
   at GoogleCloudSamples.ReidentifyContentUsingFpeTests.TestReidentifyWithFpe() in /tmpfs/src/github/dotnet-docs-samples/dlp/api/Snippets.Tests/ReidentifyContentUsingFpeTests.cs:line 30
```

**Comment by jainamtrivedi-crest** [comment]
> Tests are failing
> 
> ```cs
> Failed GoogleCloudSamples.ReidentifyContentUsingFpeTests.TestReidentifyWithFpe [3 ms]
>   Error Message:
>    System.ArgumentNullException : Value cannot be null. (Parameter 'value')
>   Stack Trace:
>      at Google.Protobuf.ProtoPreconditions.CheckNotNull[T](T value, String name)
>    at Google.Cloud.Dlp.V2.KmsWrappedCryptoKey.set_CryptoKeyName(String value)
>    at ReidentifyContentUsingFpe.ReidentifyContent(String projectId, String text, String keyName, String wrappedKey, FfxCommonNativeAlphabet alphabet, InfoType surrogateType) in /tmpfs/src/github/dotnet-docs-samples/dlp/api/Snippets/ReidentifyContentUsingFpe.cs:line 41
>    at GoogleCloudSamples.ReidentifyContentUsingFpeTests.TestReidentifyWithFpe() in /tmpfs/src/github/dotnet-docs-samples/dlp/api/Snippets.Tests/ReidentifyContentUsingFpeTests.cs:line 30
> ```

The test is failing because the code is not able to find the KeyName, which we are passing from environment variable. If you have not set it then we need to set the `KeyName` and `WrappedKey` in the environment.

P.S. README.md file has already mentioned these things; So, I have not covered it.

File: dlp/api/Snippets.Tests/ReidentifyContentUsingFpeTests.cs (hunk 1 above)
Comment (jainamtrivedi-crest) [reply dlp/api/Snippets.Tests/ReidentifyContentUsingFpeTests.cs:30]: Sure

File: dlp/api/Snippets/ReidentifyContentUsingFpe.cs (hunk 2 above)
Comment (jainamtrivedi-crest) [reply dlp/api/Snippets/ReidentifyContentUsingFpe.cs:97]: Sure

File: dlp/api/Snippets/ReidentifyContentUsingFpe.cs [hunk 4]
```diff
@@ -0,0 +1,106 @@
...
+        string keyName,
+        string wrappedKey,
+        FfxCommonNativeAlphabet alphabet = FfxCommonNativeAlphabet.Numeric,
+        InfoType surrogateType = null)
+    {
+        // Instantiate the client.
+        var dlp = DlpServiceClient.Create();
+
+        // Specify the type of info the inspection will re-identify. This must use the same custom
+        // into type that was used as a surrogate during the initial encryption.
```
[the remark this answers is not in the document]
Comment (jainamtrivedi-crest) [reply dlp/api/Snippets/ReidentifyContentUsingFpe.cs]: Done

**Comment by amanda-tarafa** [comment]
I've added the env variables. Rerunning tests...

**Review by amanda-tarafa — APPROVED** [review]

**Comment by amanda-tarafa** [comment]
Tests have now timed out. Is it possible that the encryption operations take long. @jainamtrivedi-crest can you please set a timeout for DLP tests similart to https://github.com/GoogleCloudPlatform/dotnet-docs-samples/blob/main/asset/api/runTests.ps1.
And while you are at it it would be great if you make DLP runTests.ps1 match the one linked.

Also, fix the conflicts on this PR (because I merged an earlier one).

Thanks.

**Comment by jainamtrivedi-crest** [comment]
> Tests have now timed out. Is it possible that the encryption operations take long. @jainamtrivedi-crest can you please set a timeout for DLP tests similart to https://github.com/GoogleCloudPlatform/dotnet-docs-samples/blob/main/asset/api/runTests.ps1. And while you are at it it would be great if you make DLP runTests.ps1 match the one linked.

No, Encryption does not take long time. Also, the test case has passed successfully. Not sure about the timeout error from where and why it has come. 
image

**Comment by amanda-tarafa** [comment]
With a Timeout, the test that "provokes" the timeout may very well pass. It just consumes a lot of time and then subsequent tests cannot run.

This one it's towards the higher end of running times, ~10^2ms where most seem to be ~1ms or ~10ms. There are a few others that are running in the order of seconds though.

But consuming the default timeout we have it's normal for products with plenty of samples as DLP so this is nothing to worry about. Just increasing the timeout for the test run as you've done should do the trick.

I'm rerunning tests now.

**Pull request merged**